src/message/header/mod.rs

```diff
@@ -14,3 +19,661 @@ mod date;
 mod mailbox;
 mod special;
 mod textual;
+
+pub trait Header: Clone {
+    fn name() -> HeaderName;
+
+    fn parse(s: &str) -> Result<Self, BoxError>;
+
+    fn display(&self) -> String;
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct Headers {
+    headers: Vec<(HeaderName, String)>,
+}
+
+impl Headers {
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            headers: Vec::new(),
+        }
+    }
+
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            headers: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn get<H: Header>(&self) -> Option<H> {
+        self.get_raw(&H::name()).and_then(|raw| H::parse(raw).ok())
+    }
+
+    pub fn set<H: Header>(&mut self, header: H) {
+        self.set_raw(H::name(), header.display());
+    }
+
+    pub fn remove<H: Header>(&mut self) -> Option<H> {
+        self.remove_raw(&H::name())
+            .and_then(|(_name, raw)| H::parse(&raw).ok())
+    }
+
+    #[inline]
+    pub fn clear(&mut self) {
+        self.headers.clear();
+    }
+
+    pub fn get_raw(&self, name: &str) -> Option<&str> {
+        self.find_header(name).map(|(_name, value)| value)
+    }
+
+    pub fn get_raw_mut(&mut self, name: &str) -> Option<&mut String> {
+        self.find_header_mut(name).map(|(_name, value)| value)
+    }
```

Comment:
Thinking more about it we may not want to have this be public, just in case we decide we don't want to use a `String` anymore as the backing storage.